main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -397,13 +318,132 @@ object Lexer {
+    root.put("inject", TokenKind.KeywordInject)
+    root.put("instance", TokenKind.KeywordInstance)
+    root.put("instanceof", TokenKind.KeywordInstanceOf)
+    root.put("into", TokenKind.KeywordInto)
+    root.put("law", TokenKind.KeywordLaw)
+    root.put("lawful", TokenKind.KeywordLawful)
+    root.put("lazy", TokenKind.KeywordLazy)
+    root.put("let", TokenKind.KeywordLet)
+    root.put("match", TokenKind.KeywordMatch)
+    root.put("mod", TokenKind.KeywordMod)
+    root.put("mut", TokenKind.KeywordMut)
+    root.put("new", TokenKind.KeywordNew)
+    root.put("not", TokenKind.KeywordNot)
+    root.put("null", TokenKind.KeywordNull)
+    root.put("open_variant", TokenKind.KeywordOpenVariant)
+    root.put("open_variant_as", TokenKind.KeywordOpenVariantAs)
+    root.put("or", TokenKind.KeywordOr)
+    root.put("override", TokenKind.KeywordOverride)
+    root.put("par", TokenKind.KeywordPar)
+    root.put("pquery", TokenKind.KeywordPQuery)
+    root.put("project", TokenKind.KeywordProject)
+    root.put("psolve", TokenKind.KeywordPSolve)
+    root.put("pub", TokenKind.KeywordPub)
+    root.put("query", TokenKind.KeywordQuery)
+    root.put("redef", TokenKind.KeywordRedef)
+    root.put("region", TokenKind.KeywordRegion)
+    root.put("restrictable", TokenKind.KeywordRestrictable)
+    root.put("run", TokenKind.KeywordRun)
+    root.put("rvadd", TokenKind.KeywordRvadd)
+    root.put("rvand", TokenKind.KeywordRvand)
+    root.put("rvnot", TokenKind.KeywordRvnot)
+    root.put("rvsub", TokenKind.KeywordRvsub)
+    root.put("sealed", TokenKind.KeywordSealed)
+    root.put("select", TokenKind.KeywordSelect)
+    root.put("solve", TokenKind.KeywordSolve)
+    root.put("spawn", TokenKind.KeywordSpawn)
+    root.put("static", TokenKind.KeywordStatic)
+    root.put("struct", TokenKind.KeywordStruct)
+    root.put("throw", TokenKind.KeywordThrow)
+    root.put("trait", TokenKind.KeywordTrait)
+    root.put("true", TokenKind.KeywordTrue)
+    root.put("try", TokenKind.KeywordTry)
+    root.put("type", TokenKind.KeywordType)
+    root.put("typematch", TokenKind.KeywordTypeMatch)
+    root.put("unchecked_cast", TokenKind.KeywordUncheckedCast)
+    root.put("unsafe", TokenKind.KeywordUnsafe)
+    root.put("unsafely", TokenKind.KeywordUnsafely)
+    root.put("use", TokenKind.KeywordUse)
+    root.put("where", TokenKind.KeywordWhere)
+    root.put("with", TokenKind.KeywordWith)
+    root.put("without", TokenKind.KeywordWithout)
+    root.put("xor", TokenKind.KeywordXor)
+    root.put("xvar", TokenKind.KeywordXvar)
+    root.put("yield", TokenKind.KeywordYield)
+
+    root
+  }
+
+  private def acceptKeyword()(implicit s: State): Option[TokenKind] = {
+    def advanceIfSome[T](offset: Int, opt: Option[T]): Option[T] = {
+      opt.foreach(_ => s.sc.advanceN(offset))
+      opt
+    }
+
+    @tailrec
+    def search(offset: Int, node: MutPrefixTree.Node[TokenKind]): Option[TokenKind] = {
+      s.sc.nth(offset) match {
+        case Some(c) =>
```

Comment:
For extra performance, you probably want to avoid options in the search

Reply:
I think you should keep those options everywhere-- except for where you eat one char at time.

Reply:
I suppose we can still make a clean PR and merge it. Then you can optimize in a subsequent PR.